src/AutoRest.CSharp/Mgmt/Decorator/CodeModelExtension.cs

```diff
@@ -76,17 +76,19 @@ public static void VerifyAndUpdateFrameworkTypes(this IEnumerable<Schema> allSch
                     // Do not use property.SerializedName=="type" so that we can still use x-ms-client-name to override the auto-renaming here if there is some edge case.
                     if (property.Language.Default.Name.Equals("type", StringComparison.OrdinalIgnoreCase))
                     {
+                        // TODO: we may reuse the IsResourceModel method from SchemaExtension, but the result for flattened properties is different as although models with matched flattened properties are not treated as Resource but they still inherit from ResourceData. Here we align with the inheritance match to include flattened properties for match.
                         if (property.Schema.Type == AllSchemaTypes.String
                             && objSchema.ContainsStringProperty("id")
-                            && objSchema.ContainsStringProperty("name"))
+                            && objSchema.ContainsStringProperty("name")
+                            || objSchema.Language.Default.Name.Contains("NameAvailability", StringComparison.Ordinal))
```

Comment:
Shall we add some parenthesis to emphasize the priority of this logical expression?
Or introduce another function to increase the readability here?
I believe we are changing this in this branch, if either
1. this object also has `id` and `name`, which means this is a resource model
2. this object has `NameAvailability` in it, which means this is a check name availability request
Is this correct?